### PR TITLE
[Move-compiler] introduce additional stepped compiler Translate

### PR DESCRIPTION
**Pull request by zsluedem**
## Motivation
The goal of the pr is to introduce an additional stepped compiler between parsing and expansion.

Compiler is doing from `code string -> ast -> expansion ast -> .....` to `code string -> cst -> ast -> expansion ast`.

A new cst structure is introduced which is lossless, meaning that they represent Move programs in sufficient detail to go from text to trees and vice-versa.

Incremental step on the way to https://github.com/move-language/move/issues/166

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

Yes.

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

Tests are added.

### CODE REVIEW

File: language/move-compiler/src/command_line/compiler.rs [hunk 1]
```diff
@@ -215,12 +218,14 @@ impl<'a> Compiler<'a> {
...
         let (source_text, pprog_and_comments_res) =
-            parse_program(&mut compilation_env, maps, targets, deps)?;
-        let res: Result<_, Diagnostics> = pprog_and_comments_res.and_then(|(pprog, comments)| {
-            SteppedCompiler::new_at_parser(compilation_env, pre_compiled_lib, pprog)
-                .run::<TARGET>()
-                .map(|compiler| (comments, compiler))
-        });
+            parse_program(&mut compilation_env, targets, deps)?;
+        let res: Result<_, Diagnostics> = pprog_and_comments_res
+            .and_then(|pprog| translate_program(&mut compilation_env, maps, pprog))
```
Comment (tnowacki) [inline language/move-compiler/src/command_line/compiler.rs:223]: This would need to become a new step in the stepped compiler.

File: language/move-compiler/src/parser/parsed_tree/translate/translate_address.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) The Diem Core Contributors
```
Comment (tnowacki) [inline language/move-compiler/src/parser/parsed_tree/translate/translate_address.rs:1]: I'm not a fan of these separate modules/files for each section of the ast. I know the files get very large, but having everything in one place makes understanding the entire flow easier in my opinion. (plus it is then more consistent with the rest of the compiler)

**Review by tnowacki — CHANGES_REQUESTED** [review]
Thanks for starting this! It will be a lot of work, but it will be worth it! And this looks like a good first step.  

I will try to leave some more specific comments in individual files. But some comments:

- It's parse tree or concrete syntax tree, not parse*d* tree :) 
- The ast/translate naming combo is off here. We do `<pass>::ast` and `<pass>::translate(<previous_pass>::ast::Program) -> <pass>::ast::Program`. In other words, the `translate` for a pass is the one that makes that AST. But I think things are a bit funky for the parser as it is doing a lot. I could see a few possible setups:
  - `parser/cst.rs` and `parser/ast.rs` with `parser/translate.rs` that converts the `cst` to an `ast`
  - `parser/parse_tree.rs` and `parser/ast.rs` with `parser/convert_tree.rs` that converts the `parse_tree` to an `ast`
  - `parser/cst.rs` and `<cool_new_verb>/ast.rs` with `<cool_new_verb>/translate.rs` that converts the `cst` to the `ast`
- I want to highlight again that I think all these individual sections should just be in one file. I find it very hard to follow otherwise.

File: language/move-compiler/src/parser/parsed_tree/ast.rs [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+pub type NameAccessChain = SpannedWithComment<Vec<Name>>;
+pub type LeadingNameAccess = Name;
+pub type SpecApplyFragment = ParsedToken;
+
+pub type QuantKind = SpannedWithComment<QuantKind_>;
+pub type BinOp = SpannedWithComment<BinOp_>;
+pub type UnaryOp = SpannedWithComment<UnaryOp_>;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ParsedTree {
```
Comment (tnowacki) [inline language/move-compiler/src/parser/parsed_tree/ast.rs:127]: This is going to need to be generalized a lot more (as briefly discussed in discord). The goal here is to fully capture the syntax that is in the file, with all whitespace and comment information. It can also be helpful to be able to represent invalid trees (or at least partially invalid trees). As such, parse trees typically do not use distinct types for different syntactic groups

File: language/move-compiler/src/parser/parsed_tree/translate/translate_address.rs (hunk 2 above)
Comment (zsluedem) [reply language/move-compiler/src/parser/parsed_tree/translate/translate_address.rs:1]: OK. Copy that.

Comment (zsluedem) [reply language/move-compiler/src/parser/parsed_tree/translate/translate_address.rs:1]: I was thinking I would migrate these files into one when the work is done(tests passed) because I am more comfortable with seperate files while developing. I planned to ask after the work is done. But nvm, I would definitely keep this consistent with other packages.

File: language/move-compiler/src/parser/parsed_tree/ast.rs (hunk 3 above)
Comment (zsluedem) [reply language/move-compiler/src/parser/parsed_tree/ast.rs:127]: But I was having some parsing conflicts with 

```move
spec <Identifier>{
...
}
```
 which could be parsed into `Module` representation or `SpecBlock` representation.
 
And I want to distinguish `Module` and `SpecBlock` because the items inside Module and SpecBlock are different and I want to distinguish them. So I fall back into different syntactic groups.

Hm.... Maybe I should just take `Module` and `SpecBlock` as the same enum and distinguish them later in the translation steps. 

Do you have any preferences on the point above?

**Comment by zsluedem** [comment]
@tnowacki The pr is far from review ready right now. Thank you very much for your time. It could really help keep me on the right track.

File: language/move-compiler/src/parser/parsed_tree/ast.rs (hunk 3 above)
Comment (tnowacki) [reply language/move-compiler/src/parser/parsed_tree/ast.rs:127]: You will probably have to parse it more generally, and then separate it into distinct AST nodes in this separate translation step I described above. 

In short, expect this parse tree to be _very_ general. It will more or less correspond to the structure of the current parser, but not any of the Rust code inside that structure. All the more complex/contextual analysis will need to be done in this pass that converts the parse tree to the abstract syntax tree

File: language/documentation/examples/diem-framework/move-packages/DPN/sources/AccountAdministrationScripts.move [hunk 4]
```diff
@@ -356,7 +356,7 @@ module DiemFramework::AccountAdministrationScripts {
 
         /// **Access Control:**
         /// Only the Diem Root account can process the admin scripts [[H9]][PERMISSION].
-        requires Roles::has_diem_root_role(dr_account); /// This is ensured by DiemAccount::writeset_prologue.
+        requires Roles::has_diem_root_role(dr_account); // This is ensured by DiemAccount::writeset_prologue.
```
Comment (zsluedem) [inline language/documentation/examples/diem-framework/move-packages/DPN/sources/AccountAdministrationScripts.move:359]: This was a `DocComment` which doesn't match any `DocComment` described language item. It didn't throw warning of `InvalidDocComment`. I made some of the same doc comments changes below.

File: language/documentation/examples/diem-framework/move-packages/DPN/sources/DiemAccount.move [hunk 5]
```diff
@@ -2335,7 +2335,7 @@ module DiemFramework::DiemAccount {
     spec module {} // switch documentation context back to module level
 
     /// # Access Control
-
+    ///
```
Comment (zsluedem) [inline language/documentation/examples/diem-framework/move-packages/DPN/sources/DiemAccount.move:2338]: The doc comment didn't require the consistency of the comments. I made some changes which the doc comments need to be consistent to be recognized. I got this idea because sometimes people would do comments like 
```
/// ------------------------------ line split mark -----------------------------------

/// doc comment content
fun a(){}
```

In this situation, the `line split mark` would take as doc comment with `doc comment content` which doesn't make too much sense to me.

I made the changes which the doc comment is only matched right before the language item.

Please let me know if you think this is not ok.

Comment (zsluedem) [reply language/documentation/examples/diem-framework/move-packages/DPN/sources/DiemAccount.move:2338]: A lot of comment changes in `diem-framework` are made because of it.

File: language/evm/stdlib/sources/Evm.move [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /// Module which provides access to EVM functionality, including information about the executing transaction.
-////
+///
```
Comment (zsluedem) [inline language/evm/stdlib/sources/Evm.move:2]: I don't know why but removing this could help this failing test case pass https://github.com/move-language/move/runs/8219288895?check_suite_focus=true . I don't understand how comments affect the results. It would be great someone could give me some lights.

File: language/move-compiler/src/command_line/compiler.rs [hunk 7]
```diff
@@ -214,9 +217,11 @@ impl<'a> Compiler<'a> {
...
         )?;
         let mut compilation_env = CompilationEnv::new(flags);
-        let (source_text, pprog_and_comments_res) =
-            parse_program(&mut compilation_env, maps, targets, deps)?;
-        let res: Result<_, Diagnostics> = pprog_and_comments_res.and_then(|(pprog, comments)| {
+        let (source_text, pprog_res) = parse_program(&mut compilation_env, targets, deps, maps)?;
+        let mut comments = CommentMap::new();
+
+        let res: Result<_, Diagnostics> = pprog_res.and_then(|pprog| {
+            find_comments(&pprog, &mut compilation_env, &mut comments)?;
```
Comment (zsluedem) [inline language/move-compiler/src/command_line/compiler.rs:224]: Maybe this is a `TODO`. It would be greate to move this only to `doc-gen` command.

File: language/move-compiler/tests/move_check/flavors/async/derive_errors.exp [hunk 8]
```diff
@@ -69,12 +69,6 @@ error[E13001]: attribute derivation failed
 22 │     #[message] fun f7(_x: &A, _a1: u64, _a2: u64, _a3: u64, _a4: u64, _a5: u64, _a6: u64,
    │       ^^^^^^^ too many arguments for message handler (current max is 8)
 
-error[E03003]: unbound module member
-   ┌─ tests/move_check/flavors/async/derive_errors.move:22:7
-   │
-22 │     #[message] fun f7(_x: &A, _a1: u64, _a2: u64, _a3: u64, _a4: u64, _a5: u64, _a6: u64,
-   │       ^^^^^^^ Invalid module access. Unbound function 'send__9' in module '(Async=0x20)::Runtime'
-
```
Comment (zsluedem) [inline language/move-compiler/tests/move_check/flavors/async/derive_errors.exp:77]: This error message is gone because I moved the checking into any layer which is `Expansion`. The derived check failed first but the unbound check haven't started.

File: language/move-compiler/tests/move_check/parser/spec_parsing_inside_fun.exp [hunk 9]
```diff
@@ -1,54 +1,9 @@
...
-   │
-34 │         &mut spec {};
-   │         ^^^^^^^^^^^^
-   │         │    │
-   │         │    Expected a single non-reference type, but found: '()'
-   │         Invalid borrow
+   │                 ^
+   │                 │
+   │                 Unexpected '+'
+   │                 Expected an expression term
```
Comment (zsluedem) [inline language/move-compiler/tests/move_check/parser/spec_parsing_inside_fun.exp:8]: I am actually a little bit confused that why a `Spec` could be a member of `Expression` https://github.com/move-language/move/blob/164bf33c8b396b9a29305d9c1b9a0d3b0c5b8cef/language/move-compiler/src/parser/ast.rs#L643-L644. Maybe I missed something that is important. Please let me know.

I chose to let `cst` behaved similarly with `ast` which `Exp` includes `Spec` to make the test https://github.com/move-language/move/blob/164bf33c8b396b9a29305d9c1b9a0d3b0c5b8cef/language/move-compiler/tests/move_check/typing/spec_block_fail.move#L1-L7 pass.

I'd like to let this fail earlier in the parsing which `spec` doesn't work like `Exp` . 

Please let me know your ideas on it.

File: language/move-compiler/src/parser/parsed_tree/ast.rs (hunk 3 above)
Comment (zsluedem) [reply language/move-compiler/src/parser/parsed_tree/ast.rs]: @tnowacki Please take a look whether the pr is on the right track or not~!

**Comment by zsluedem** [comment]
Got other plans

**Comment by yuyang-ok** [comment]
@zsluedem What's you new plans??? :-)

**Comment by zsluedem** [comment]
@yuyang-ok Well. I got a crazy idea to rewrite the parser in another repo and probably rewrite the vs code plugin.

The parser would parse the code into a red-green tree ( lossless syntax trees).
The parser should also do the incremental reparsing  easily which the vscode plugin would benefit a lot.
